### PR TITLE
Documentation and cleanup of automatic discovery of inverse associations

### PR DESCRIPTION
**Pull request by wangjohn**
Removed `:automatic_inverse_of` in favor of using the `inverse_of: false` to tell us whether we don't want to use automatic inverses. Changing the documentation and adding a CHANGELOG entry for the automatic inverse detection feature.

@jonleighton I don't think it's actually possible to use `inverse_of: nil` to tell us to not automatically find inverse associations. This is because the feature would be completely useless, since by default, we have `options[:inverse_of] = nil`. This means, by default we do not use automatic inverse finding.

Conversely, having the user specify `options[:inverse_of] = true` to use automatic associations doesn't make sense because at that point, the user can make the extra couple characters and type in the inverse.

### CODE REVIEW

**Comment by duduribeiro** [comment]
:+1:

**Comment by cyrilchampier** [comment]
sorry if this is not the right place to ask, but what would be a good reason to disable this behavior ?

File: activerecord/lib/active_record/reflection.rb [hunk 1]
```diff
@@ -449,15 +448,16 @@ def valid_inverse_reflection?(reflection)
...
+        # have <tt>has_many</tt>, <tt>has_one</tt>, <tt>belongs_to</tt> associations.
+        # Third, we must not have options such as <tt>:polymorphic</tt> or
+        # <tt>:foreign_key</tt> which prevent us from correctly guessing the
+        # inverse association.
         #
         # Anything with a scope can additionally ruin our attempt at finding an
         # inverse, so we exclude reflections with scopes.
         def can_find_inverse_of_automatically?(reflection)
-          reflection.options[:automatic_inverse_of] != false &&
+          reflection.options[:inverse_of] != false &&
```
[the remark this answers is not in the document]
Comment (Victorcorcos) [reply activerecord/lib/active_record/reflection.rb]: Very interesting line of reasoning.

[the remark this answers is not in the document]
Comment (Victorcorcos) [reply activerecord/lib/active_record/reflection.rb]: @egilburg Very interesting line of reasoning.

**Pull request merged**